src/memory.c

```diff
@@ -149,87 +132,112 @@ static int pagesize;
 {
   for (int i = 0; i < ci->children_num; i++) {
     oconfig_item_t *child = ci->children + i;
-    if (strcasecmp("ValuesAbsolute", child->key) == 0)
-      cf_util_get_boolean(child, &values_absolute);
-    else if (strcasecmp("ValuesPercentage", child->key) == 0)
-      cf_util_get_boolean(child, &values_percentage);
+    if (strcasecmp("ReportUsage", child->key) == 0 ||
+        strcasecmp("ValuesAbsolute", child->key) == 0)
+      cf_util_get_boolean(child, &report_usage);
+    else if (strcasecmp("ReportUtilization", child->key) == 0 ||
+             strcasecmp("ValuesPercentage", child->key) == 0)
+      cf_util_get_boolean(child, &report_utilization);
+    else if (strcasecmp("ReportLimit", child->key) == 0)
+      cf_util_get_boolean(child, &report_limit);
     else
       ERROR("memory plugin: Invalid configuration option: \"%s\".", child->key);
   }
 
   return 0;
 } /* }}} int memory_config */
 
-static int memory_dispatch(gauge_t values[COLLECTD_MEMORY_TYPE_MAX]) {
-  metric_family_t fam_absolute = {
+static int memory_dispatch(gauge_t values[STATE_MAX]) {
+  metric_family_t fam_usage = {
       .name = "system.memory.usage",
       .help = "Reports memory in use by state",
       .unit = "By",
       .type = METRIC_TYPE_GAUGE,
   };
   gauge_t total = 0;
 
-  for (size_t i = 0; i < COLLECTD_MEMORY_TYPE_MAX; i++) {
+  for (size_t i = 0; i < STATE_MAX; i++) {
     if (isnan(values[i])) {
       continue;
     }
 
     total += values[i];
 
-    if (values_absolute) {
-      metric_family_append(&fam_absolute, label_state, memory_type_names[i],
+    if (report_usage) {
+      metric_family_append(&fam_usage, label_state, memory_type_names[i],
                            (value_t){.gauge = values[i]}, NULL);
     }
   }
 
   int ret = 0;
-  if (values_absolute) {
-    int status = plugin_dispatch_metric_family(&fam_absolute);
+  if (report_usage) {
+    int status = plugin_dispatch_metric_family(&fam_usage);
     if (status != 0) {
       ERROR("memory plugin: plugin_dispatch_metric_family failed: %s",
             STRERROR(status));
     }
     ret = status;
   }
-  metric_family_metric_reset(&fam_absolute);
-
-  if (!values_percentage) {
-    return ret;
-  }
+  metric_family_metric_reset(&fam_usage);
 
   if (total == 0) {
     return EINVAL;
   }
 
-  metric_family_t fam_percent = {
+  if (report_limit) {
+    metric_family_t fam_limit = {
+        .name = "system.memory.limit",
+        .help = "Total memory available in the system.",
+        .unit = "By",
+        .type = METRIC_TYPE_COUNTER, // [sic] should be UpDownCounter
+    };
+    metric_t m = {
+        .value = (value_t){.derive = (derive_t)total},
+    };
+    metric_family_metric_append(&fam_limit, m);
+
+    int status = plugin_dispatch_metric_family(&fam_limit);
+    if (status != 0) {
+      ERROR("memory plugin: plugin_dispatch_metric_family failed: %s",
+            STRERROR(status));
+    }
+    ret = ret ? ret : status;
+  }
```

Comment:
Installed memory amount is more like cluster node property, rather than its metric, because it cannot change (without taking the node down).  If usage + utilization ratio are not enough, report free memory instead (user can then derive total from usage+free, if s/he wants).

Reply:
I think the target audience here are Cloud workloads, e.g. containers. The runtime may do automatic resizing and the question "how much memory is allocated to my service" is both interesting and not static.

For physical machines and VMs this is not very interesting, and so I think the default of disabling the reporting of this metric is correct.

Reply:
> I think the target audience here are Cloud workloads, e.g. containers. The runtime may do automatic resizing and the question "how much memory is allocated to my service" is both interesting and not static.

Sure, but that's not what the plugin provides.

Containers' memory usage is limited by memory cgroup: https://docs.kernel.org/admin-guide/cgroup-v2.html#memory

Whereas `/proc/meminfo` read by the memory plugin lists the whole system memory statistics, which total does **not** change.

Reply:
That's a good point. I also agree that this is not really a "metric" – I have argued against the *uptime plugin* along similar lines. So if you feel strongly about this, we can revert 413112fe93888e13eb8b4c11fc09ecd1159e710f.

On the other hand, many people don't have an inventory system alongside their metrics, and fewer yet can merge data between the two. I think there are situations where having the number of CPUs, the amount of memory, etc. available *as a metric* can be useful for some people. With the metric being disabled by default, I don't think we negatively impact users that don't need it. So if you don't feel strongly about this, I'd say let's keep the metric, it doesn't impact the user experience unless activated.

Reply:
Good point about inventory.  Do not feel strongly about this / it's OK being disabled by default.